mega.py

```diff
@@ -0,0 +1,85 @@
+import numpy as np
+import random
+
+from arms import ArmBernoulli
+from users import SecondaryUser
+
+from plots import regret_plt
+
+#  UNIVERSE PARAMETERS
+n_users = 3
+n_arms = 5
+t_horizon = 100000
+
+arm_means = [0.2, 0.3, 0.5, 0.8, 0.9]
+
+arms = list()
+for i in range(n_arms):
+    arms.append(ArmBernoulli(arm_means[i]))
+
+# ALGORITHM PARAMETERS
+params = {
+    'c': 0.1,
+    'd': 0.05,
+    'alpha': 0.5,
+    'beta': 0.8,
+    'persistence_proba_init': 0.6
+}
+
+users = list()
+for i in range(n_users):
+    users.append(SecondaryUser(n_arms, params))
+
+# global statistics
+rewards = np.zeros((n_users, t_horizon))
+collisions = np.zeros((n_users, t_horizon), dtype=np.uint8)
+occupation = np.zeros(n_users)
+# main loop
+for t in range(1, t_horizon):
+    # update collision status
+    for i, user in enumerate(users):
+        if not user.collided:
+            # increase its persistence probability
+            user.persistence_proba *= user.params['alpha']
+            user.persistence_proba += user.params['alpha']
+        else:
+            collisions[i, t-1] = 1
+            if random.random() < user.persistence_proba:
+                # the users persists !
+                user.previous_arm = user.arm
```

Comment:
C'est un peu tordu. Ces variables contiennent le bras *arm* et le précédent *previous_arm* au temps *t*. Or à ce moment de la boucle elles contiennent encore les valeurs de l'itération précédente. 

*previous_arm* contient le bras à t-2 donc on peut l'effacer.
Et *arm* contient le bras à t-1.
Du coup on fait previous_arm <- arm et *arm* ne change pas car l'utilisateur persiste !